frontend/apps/reader/modules/readergoto.lua

```diff
@@ -181,4 +181,28 @@ function ReaderGoto:onGoToEnd()
     return true
 end
 
+function ReaderGoto:onGoToRandomPage()
+    local page_count = self.document:getPageCount()
+    if page_count == 1 then return true end
+    local current_page = self.ui:getCurrentPage()
+    if self.pages_pool == nil then
+        self.pages_pool = {}
+    end
+    if #self.pages_pool == 0 or (#self.pages_pool == 1 and self.pages_pool[1] == current_page) then
+        for i = 1, page_count do
+            self.pages_pool[i] = i
+        end
+    end
+    while true do
+        local random_page_idx = math.random(1, #self.pages_pool)
+        local random_page = self.pages_pool[random_page_idx]
+        if random_page ~= current_page then
+            table.remove(self.pages_pool, random_page_idx)
+            self.ui.link:addCurrentLocationToStack()
+            self.ui:handleEvent(Event:new("GotoPage", random_page))
+            return true
+        end
+    end
```

Comment:
I would do the inverse.
Instead of having a table with page_num entries always (wasting memory), i would start an empty table and add the new random page to it, then the loop becomes `while self.pages_pool[new_random_page] do` ...(get new random page)

Reply:
I tested with a 36 page PDF, the initial page number was 9, I opened the file and jumped to random pages 31 times, then closed the file and opened another file and swiped a few times, then returned to the original file, pages 4, 9, 18, 22, 28, 29, 30 are displayed twice.

If you don't have multiple files open at the same time, the page_pool entries won't waste too much memory.

<details>
<summary>Test data for <b>Random Page</b></summary>

<b>9</b>, the initial page
14
25
8
21
19
5
31
29
12
2
10
17
28
22
9
18
15
23
24
3
36
30
6
20
13
27
4
26
16
32
7, I close the file here, open another file swipe several times, then open the original file
<b>30
28
22
4
29
18</b>

</details>

In a file of more than 200 or 1000 page, maybe we should generate random number without page_pool:

```lua
    if page_count > 200 then
        local random_page = math.random(1, page_count)
        self.ui.link:addCurrentLocationToStack()
        self.ui:handleEvent(Event:new("GotoPage", random_page))
        return true
    else
        # generate random page from page_pool
```

Reply:
> i would start an empty table and add the new random page to it

You cannot guarantee that the next random draw gives unvisited page.
**Theoretically**, you can get an infinite loop.

Reply:
> then closed the file and opened another file

Once you close the file, the visited pages record is lost.

Reply:
>Once you close the file, the visited pages record is lost.

This means memory will not be wasted when we open at most 2 files at the same time.

For memory sensitive users, we can generate random page without using page_pool for books with more than N pages.